src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs

```diff
@@ -30,31 +29,75 @@ namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
+                    WithCommitCharacterRule(CharacterSetModificationRule.Create(CharacterSetModificationKind.Add, '<'));
+
+                var importDirectiveBaseline = CompletionItemRules.Create(commitCharacterRules:
+                    ImmutableArray.Create(CharacterSetModificationRule.Create(CharacterSetModificationKind.Replace, '.', ';')));
+
+                var rule = context.importDirective ? importDirectiveBaseline : generalBaseline;
+
+                if (context.preselect)
+                    rule = rule.WithSelectionBehavior(CompletionItemSelectionBehavior.HardSelection);
+
+                if (context.tupleLiteral)
+                    rule = rule.WithCommitCharacterRule(CharacterSetModificationRule.Create(CharacterSetModificationKind.Remove, ':'));
+
+                return rule;
+            }
+        }
+
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public SymbolCompletionProvider()
         {
         }
 
-        protected override Task<ImmutableArray<ISymbol>> GetSymbolsAsync(SyntaxContext context, int position, OptionSet options, CancellationToken cancellationToken)
-        {
-            return Recommender.GetImmutableRecommendedSymbolsAtPositionAsync(
-                context.SemanticModel, position, context.Workspace, options, cancellationToken);
-        }
+        protected override CompletionItemSelectionBehavior PreselectedItemSelectionBehavior => CompletionItemSelectionBehavior.HardSelection;
 
-        protected override async Task<bool> ShouldProvidePreselectedItemsAsync(CompletionContext completionContext, SyntaxContext syntaxContext, Document document, int position, OptionSet options)
+        protected override async Task<bool> ShouldPreselectInferredTypesAsync(
+            CompletionContext? context,
+            int position,
+            OptionSet options,
+            CancellationToken cancellationToken)
         {
-            var sourceText = await document.GetTextAsync(CancellationToken.None).ConfigureAwait(false);
-            if (ShouldTriggerInArgumentLists(sourceText, options))
+            if (context != null)
             {
-                // Avoid preselection & hard selection when triggered via insertion in an argument list.
-                // If an item is hard selected, then a user trying to type MethodCall() will get
-                // MethodCall(someVariable) instead. We need only soft selected items to prevent this.
-                if (completionContext.Trigger.Kind == CompletionTriggerKind.Insertion &&
-                    position > 0 &&
-                    await IsTriggerInArgumentListAsync(document, position - 1, CancellationToken.None).ConfigureAwait(false) == true)
+                var document = context.Document;
+                var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
we weren't passing along cancellation :-/

Reply:
We shouldn't need `sourceText` at all. The only thing it's used for is getting `IExperimentationService`, and we can get that directly from `document`.

Reply:
yup.  got rid of the need for this.